File: backend/internal/file/deletion_worker.go
package file

import (
	"context"
	"log"
	"time"
)

// DeletionWorker 在数据库事务提交后异步清理对象，任务本身持久化在 MySQL。
type DeletionWorker struct {
	queue  ObjectDeletionQueue
	delete func(context.Context, string) error
}

func NewDeletionWorker(queue ObjectDeletionQueue, deleteObject func(context.Context, string) error) *DeletionWorker {
	return &DeletionWorker{queue: queue, delete: deleteObject}
}

func (w *DeletionWorker) Run(ctx context.Context) {
	if err := w.queue.Ensure(ctx); err != nil {
		log.Printf("object deletion outbox unavailable: %v", err)
		return
	}
	ticker := time.NewTicker(2 * time.Second)
	defer ticker.Stop()
	for {
		w.drain(ctx)
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
		}
	}
}

func (w *DeletionWorker) drain(ctx context.Context) {
	tasks, err := w.queue.Pending(ctx, 20)
	if err != nil {
		log.Printf("object deletion outbox read failed: %v", err)
		return
	}
	for _, task := range tasks {
		if err := w.delete(ctx, task.ObjectKey); err != nil {
			if retryErr := w.queue.Retry(ctx, task.ID, 1); retryErr != nil {
				log.Printf("object deletion retry update failed for task %d: %v", task.ID, retryErr)
			}
			continue
		}
		if err := w.queue.Complete(ctx, task.ID); err != nil {
			log.Printf("object deletion task completion failed for task %d: %v", task.ID, err)
		}
	}
}

File: backend-java/src/main/java/com/clouddrive/file/DeletionWorker.java
package com.clouddrive.file;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.scheduling.annotation.Scheduled;

/**
 * 对象删除 worker，对应 Go file.DeletionWorker。每 2 秒清理一批（最多 20），失败指数退避。
 */
@org.springframework.stereotype.Component
public class DeletionWorker {

	private static final Logger log = LoggerFactory.getLogger(DeletionWorker.class);

	private final ObjectDeletionQueue queue;

	private final ObjectStore objects;

	public DeletionWorker(ObjectDeletionQueue queue, ObjectStore objects) {
		this.queue = queue;
		this.objects = objects;
	}

	@org.springframework.beans.factory.annotation.Value("${deletion.worker.enabled:true}")
	private boolean enabled;

	@Scheduled(fixedDelay = 2000L)
	public void drain() {
		if (!enabled) {
			return;
		}
		try {
			for (DeletionTask task : queue.pending(20)) {
				try {
					objects.delete(task.objectKey());
					queue.complete(task.id());
				}
				catch (RuntimeException e) {
					log.warn("object deletion retry for task {}: {}", task.id(), e.getMessage());
					try {
						queue.retry(task.id(), 1);
					}
					catch (RuntimeException retryErr) {
						log.warn("object deletion retry update failed for task {}: {}", task.id(),
								retryErr.getMessage());
					}
				}
			}
		}
		catch (RuntimeException e) {
			log.warn("object deletion outbox read failed: {}", e.getMessage());
		}
	}

}